src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/DataSource.scala

```diff
@@ -112,43 +113,103 @@ class DataSource(graphFactory: OrientGraphFactory)(implicit executionContext: Ex
   def shutdown() = graphFactory.close()
 }
 
-class RawlsTransaction(val graph: OrientGraph, dataSource: DataSource) {
-  private val rollbackOnly = new AtomicBoolean(false)
-
+object RawlsTransaction {
   /**
    * We have to use JVM locks because Orient refuses to lock remote databases
    */
-  var elemLocks = TrieMap[Element, ReentrantReadWriteLock]()
+  var wsLocks = TrieMap[WorkspaceName, StampedLock]()
 
-  private def getElemLock(elem: Element): ReentrantReadWriteLock = {
-    elemLocks.getOrElseUpdate(elem, new ReentrantReadWriteLock())
+  private def getLock(ws: WorkspaceName): StampedLock = {
+    wsLocks.getOrElseUpdate(ws, new StampedLock())
   }
+}
 
-  def withWriteLock[T](elem: Element) (op: => T): T = {
-    val wl = getElemLock(elem).writeLock()
-    try {
-      wl.lock()
-      op
-    } finally {
-      wl.unlock()
+class RawlsTransaction(val graph: OrientGraph, dataSource: DataSource)(implicit executionContext: ExecutionContext) {
+  private val rollbackOnly = new AtomicBoolean(false)
+
+  //Count of locks held by this txn. Handles silently upgrading read to write locks if a write is already held.
+  private val readCounts = TrieMap[WorkspaceName, AtomicInteger]().withDefaultValue(new AtomicInteger(0))
+  private val writeCounts =  TrieMap[WorkspaceName, AtomicInteger]().withDefaultValue(new AtomicInteger(0))
+
+  private def acquireReadLock(ws: WorkspaceName) = {
+    if( writeCounts(ws).get > 0 ) {
+      //Silently "upgrade" read locks to write locks if a write lock exists, since we already
+      //have exclusive access.
+      acquireWriteLock(ws)
+    } else if( readCounts(ws).getAndIncrement == 0 ) {
+      RawlsTransaction.getLock(ws).readLock()
     }
   }
 
-  def withReadLock[T](elem: Element) (op: => T): T = {
-    val rl = getElemLock(elem).readLock()
-    try {
-      rl.lock()
-      op
-    } finally {
-      rl.unlock()
+  private def releaseReadLock(ws: WorkspaceName) = {
+    if( writeCounts(ws).get > 0 ) {
+      //Handle our read lock having been silently upgraded.
+      releaseWriteLock(ws)
+    } else {
+      assert(readCounts(ws).get > 0, s"Attempting to release nonexistent read lock on $ws")
+      if( readCounts(ws).decrementAndGet == 0 ) {
+        val unlocked = RawlsTransaction.getLock(ws).tryUnlockRead()
+        assert(unlocked, s"Someone else apparently released read lock on $ws ahead of readCount == 0???")
```

Comment:
`tryUnlockRead` attempts to release one hold of the read lock and returns whether it did so. If there was no read lock, something's gone wrong, because this thread should have been the one that had it?

This assert is essentially "shouldn't get here". That sentence makes me nervous.